File: packages/sdk-python/amanai/operators.py
"""Operator match logic — the single source of truth for rule predicates.

Runtime enforcement (`policy.py`) and the offline judge (`judge.py`) both call
`op_match` here, so "what you test" and "what you enforce" can never drift —
there is one implementation, not two to keep in sync.

A predicate is *true* when the action's value satisfies the operator against the
rule's target. A true predicate means the rule applies (its action fires).
"""

from __future__ import annotations

import ipaddress
import re
from urllib.parse import urlsplit

# Every operator the policy DSL understands. Loading a rule with an op outside
# this set fails fast (see `policy.load_policy`).
KNOWN_OPERATORS = frozenset(
    {
        ">=",
        ">",
        "<=",
        "<",
        "==",
        "eq",
        "!=",
        "ne",
        "contains",
        "regex",
        "in",
        "membership",
        "not_in",
        "email_external",
        "domain_in",
        "domain_not_in",
        "internal_url",
    }
)


def _domain(value) -> str:
    return str(value).split("@")[-1].lower()


def _domains(target) -> list[str]:
    return [str(d).lower() for d in (target or [])]


# Hostnames that are internal but aren't literal IPs, so `ipaddress` can't flag
# them. `localhost` resolves to loopback; `metadata.google.internal` is the GCP
# metadata endpoint (AWS/Azure use 169.254.169.254, caught as link-local).
_INTERNAL_HOSTS = frozenset({"localhost", "metadata.google.internal"})


def _is_internal_url(value) -> bool:
    """True if `value` (a URL or bare host) points somewhere internal: a private,
    loopback, link-local, reserved, or unspecified IP, or a known-internal host.

    ponytail: no DNS resolution. Catches literal-IP and known-hostname SSRF only,
    not DNS-rebinding (a public hostname that resolves to an internal IP at
    request time). Resolving needs network + is TOCTOU-prone — out of scope for a
    deterministic in-process check. Same posture as the regex-ReDoS note below.
    """
    # urlsplit needs a scheme to populate .hostname; fall back to the raw string
    # for a bare host like "10.0.0.1" or "localhost".
    host = urlsplit(str(value)).hostname or str(value).strip()
    host = host.strip().rstrip(".").lower()  # drop FQDN root dot
    if not host:
        return False
    if host in _INTERNAL_HOSTS:
        return True
    try:
        ip = ipaddress.ip_address(host)
    except ValueError:
        return False  # a hostname we don't resolve — treat as external
    return (
        ip.is_private
        or ip.is_loopback
        or ip.is_link_local
        or ip.is_reserved
        or ip.is_unspecified
    )


# Canonical numeric grammar, shared with the TypeScript engine (operators.ts
# `pyFloat`). A security decision must coerce numbers identically in both SDKs, so
# we do NOT defer to each language's native parser: Python `float()` accepts "9_0"
# while JS `Number()` rejects it, and JS accepts "0x1f" while Python rejects it —
# that drift flips block/allow across languages. One strict grammar removes it:
# optional sign, decimal or scientific notation, plus inf/infinity/nan
# (case-insensitive). No underscores, no hex/binary/octal.
_NUM_RE = re.compile(r"[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?$")


def _to_number(x) -> float:
    """Coerce to float under the canonical grammar, else raise ValueError. bool
    coerces (True → 1.0), matching Python's own numeric `==` semantics."""
    if isinstance(x, bool):
        return float(x)
    if isinstance(x, (int, float)):
        return float(x)
    if isinstance(x, str):
        t = x.strip()
        low = t.lower().lstrip("+-")
        if low in ("inf", "infinity"):
            return float("-inf") if t.lstrip().startswith("-") else float("inf")
        if low == "nan":
            return float("nan")
        if _NUM_RE.match(t):
            return float(t)
    raise ValueError(f"not a canonical number: {x!r}")


def op_match(op: str, value, target) -> bool:
    """True when `value` satisfies `op` against `target`.

    Returns False for unknown operators or type errors (never raises) so a
    malformed value can't crash a security decision.
    """
    try:
        if op in (">=", ">", "<=", "<"):
            x, y = _to_number(value), _to_number(target)
            return {">=": x >= y, ">": x > y, "<=": x <= y, "<": x < y}[op]
        if op in ("==", "eq"):
            # Numeric-first so "5" == 5. Note: bool coerces (True == 1), matching
            # Python's own semantics; use a string value to compare literally.
            try:
                return _to_number(value) == _to_number(target)
            except (TypeError, ValueError):
                return value == target
        if op in ("!=", "ne"):
            try:
                return _to_number(value) != _to_number(target)
            except (TypeError, ValueError):
                return value != target
        if op == "contains":
            return str(target).lower() in str(value).lower()
        if op == "regex":
            # ponytail: policy-supplied regex, no ReDoS guard. Fine while policy
            # authorship is trusted; add a re2/timeout wrapper if policies ever
            # come from untrusted input.
            return re.search(str(target), str(value)) is not None
        if op in ("in", "membership"):
            return value in (target or [])
        if op == "not_in":
            return value not in (target or [])
        if op == "email_external":
            return _domain(value) not in _domains(target)
        if op == "domain_in":
            return _domain(value) in _domains(target)
        if op == "domain_not_in":
            return _domain(value) not in _domains(target)
        if op == "internal_url":
            # target is the rule's `value`: true = fire on internal, false on
            # external. Any parse error falls through to the except → False.
            return _is_internal_url(value) == bool(target)
    except Exception:
        return False
    return False

File: packages/sdk-node/src/operators.ts
/**
 * Operator match logic — the single source of truth for rule predicates, ported
 * to match Python `amanai/operators.py::op_match` exactly. Runtime enforcement
 * and the parity vectors both rely on identical semantics, so "what you test" and
 * "what you enforce" can never drift across languages.
 *
 * Like the Python version, `opMatch` never throws: an unknown operator or a type
 * error returns false so a malformed value can't crash a security decision.
 */

/** Every operator the DSL understands. An op outside this set fails policy load. */
export const KNOWN_OPERATORS = new Set<string>([
  ">=", ">", "<=", "<",
  "==", "eq", "!=", "ne",
  "contains", "regex",
  "in", "membership", "not_in",
  "email_external", "domain_in", "domain_not_in", "internal_url",
]);

/** Mirror Python `float(x)`: True→1, numeric string→number, "" / None / list → NaN. */
function pyFloat(x: unknown): number {
  if (typeof x === "number") return x;
  if (typeof x === "boolean") return x ? 1 : 0;
  if (typeof x === "string") {
    const t = x.trim();
    if (t === "") return NaN;
    return Number(t);
  }
  return NaN; // null/undefined/object → Python float() raises → treated as no-match
}

/** Numeric-first equality with a Python-like fallback (so "5" == 5). */
function pyEq(value: unknown, target: unknown): boolean {
  const a = pyFloat(value);
  const b = pyFloat(target);
  if (!Number.isNaN(a) && !Number.isNaN(b)) return a === b;
  if (typeof value === "object" || typeof target === "object") {
    return JSON.stringify(value) === JSON.stringify(target);
  }
  return value === target;
}

function asArray(target: unknown): unknown[] {
  return Array.isArray(target) ? target : [];
}

function domainOf(value: unknown): string {
  const s = String(value);
  return s.slice(s.lastIndexOf("@") + 1).toLowerCase();
}

function domains(target: unknown): string[] {
  return asArray(target).map((d) => String(d).toLowerCase());
}

// Internal hosts that aren't literal IPs (localhost → loopback; GCP metadata host).
const INTERNAL_HOSTS = new Set(["localhost", "metadata.google.internal"]);

function parseIpv4(host: string): number[] | null {
  const parts = host.split(".");
  if (parts.length !== 4) return null;
  const octets: number[] = [];
  for (const p of parts) {
    if (!/^\d{1,3}$/.test(p)) return null;
    const n = Number(p);
    if (n > 255) return null;
    octets.push(n);
  }
  return octets;
}

/**
 * True for private/loopback/link-local/reserved/unspecified addresses and known
 * internal hosts. Literal-IP + known-hostname only — no DNS resolution, matching
 * the Python posture (catches direct-IP SSRF, not DNS-rebinding).
 */
function isInternalUrl(value: unknown): boolean {
  let host: string;
  try {
    host = new URL(String(value)).hostname || String(value).trim();
  } catch {
    host = String(value).trim();
  }
  host = host.trim().replace(/\.$/, "").replace(/^\[|\]$/g, "").toLowerCase();
  if (!host) return false;
  if (INTERNAL_HOSTS.has(host)) return true;

  const v4 = parseIpv4(host);
  if (v4) {
    const [a, b] = v4;
    if (a === 10) return true; // private 10/8
    if (a === 172 && b >= 16 && b <= 31) return true; // private 172.16/12
    if (a === 192 && b === 168) return true; // private 192.168/16
    if (a === 127) return true; // loopback 127/8
    if (a === 169 && b === 254) return true; // link-local 169.254/16
    if (a === 0) return true; // unspecified / this-network
    if (a >= 240) return true; // reserved 240/4
    return false;
  }
  // Minimal IPv6 specials (full parity for v6 is out of scope — v4 + hosts only).
  if (host === "::1") return true; // loopback
  if (host.startsWith("fe80:") || host.startsWith("fc") || host.startsWith("fd")) return true;
  if (host === "::") return true; // unspecified
  return false; // a hostname we don't resolve → external
}

/** True when `value` satisfies `op` against `target`. Never throws. */
export function opMatch(op: string, value: unknown, target: unknown): boolean {
  try {
    switch (op) {
      case ">=":
      case ">":
      case "<=":
      case "<": {
        const x = pyFloat(value);
        const y = pyFloat(target);
        if (Number.isNaN(x) || Number.isNaN(y)) return false;
        return op === ">=" ? x >= y : op === ">" ? x > y : op === "<=" ? x <= y : x < y;
      }
      case "==":
      case "eq":
        return pyEq(value, target);
      case "!=":
      case "ne":
        return !pyEq(value, target);
      case "contains":
        return String(value).toLowerCase().includes(String(target).toLowerCase());
      case "regex":
        // ponytail: policy-supplied regex, no ReDoS guard — same posture as Python.
        return new RegExp(String(target)).test(String(value));
      case "in":
      case "membership":
        return asArray(target).includes(value);
      case "not_in":
        return !asArray(target).includes(value);
      case "email_external":
        return !domains(target).includes(domainOf(value));
      case "domain_in":
        return domains(target).includes(domainOf(value));
      case "domain_not_in":
        return !domains(target).includes(domainOf(value));
      case "internal_url":
        return isInternalUrl(value) === Boolean(target);
      default:
        return false;
    }
  } catch {
    return false;
  }
}
